source/extensions/transport_sockets/alts/tsi_frame_protector.cc

```diff
@@ -0,0 +1,79 @@
+#include "extensions/transport_sockets/alts/tsi_frame_protector.h"
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace TransportSockets {
+namespace Alts {
+
+TsiFrameProtector::TsiFrameProtector(CFrameProtectorPtr&& frame_protector)
+    : frame_protector_(std::move(frame_protector)) {}
+
+tsi_result TsiFrameProtector::protect(Buffer::Instance& input, Buffer::Instance& output) {
+  ASSERT(frame_protector_);
+
+  // TODO(lizan): tune size later
+  unsigned char protected_buffer[16384];
+  size_t protected_buffer_size = sizeof(protected_buffer);
+  while (input.length() > 0) {
+    auto* message_bytes = reinterpret_cast<unsigned char*>(input.linearize(input.length()));
+    size_t protected_buffer_size_to_send = protected_buffer_size;
+    size_t processed_message_size = input.length();
+    tsi_result result =
+        tsi_frame_protector_protect(frame_protector_.get(), message_bytes, &processed_message_size,
+                                    protected_buffer, &protected_buffer_size_to_send);
+    if (result != TSI_OK) {
+      ASSERT(result != TSI_INVALID_ARGUMENT && result != TSI_UNIMPLEMENTED);
+      return result;
+    }
+    output.add(protected_buffer, protected_buffer_size_to_send);
+    input.drain(processed_message_size);
+  }
+
+  ASSERT(input.length() == 0);
```

Comment:
Is this ASSERT redundant as the while condition guarantees input.length()==0 to exit the loop?